### ISSUE TITLE
How does horovod coordinates training?

### ISSUE THREAD

**Issue by josesuero** [horovod/horovod#820]
I'm trying to learn how to train a distributed Keras model and found horovod, I created a Google Cloud Kubernetes cluster (GKE) and installed Kubeflow with the mpi-operator and the nvdia drivers.

I tried running the sample https://github.com/uber/horovod/blob/master/examples/keras_mnist_advanced.py

I created a Docker image with this definition:

```
FROM uber/horovod:0.15.2-tf1.12.0-torch1.0.0-py3.5

# Temporary fix until Horovod pushes out a new release.
# See https://github.com/uber/horovod/pull/700
RUN sed -i '/^NCCL_SOCKET_IFNAME.*/d' /etc/nccl.conf


RUN mkdir /home/mnist
WORKDIR /home/mnist
ADD keras_mnist_advanced.py /home/mnist

RUN ldconfig /usr/local/cuda/lib64/stubs

CMD mpirun \
 python keras_mnist_advanced.py
```
pushed it to the project repository and made it work, with the following Yaml

```
apiVersion: kubeflow.org/v1alpha1
kind: MPIJob
metadata:
  name: keras-mnist-1
spec:
  replicas: 3
  template:
    spec:
      containers:
      - image: us.gcr.io/$ProjectName/keras-mnist:1.0
        name: keras-mnist
```
This runs 3 workers, and a launcher, saves checkpoints only on worker-0, all good, but the logs really confuse me,

I get 3 lines of logs (one per worker)

```
I  x_train shape: (60000, 28, 28, 1)

I  60000 train samples

I  10000 test samples

I  x_train shape: (60000, 28, 28, 1)

I  60000 train samples

I  10000 test samples

I  x_train shape: (60000, 28, 28, 1)

I  60000 train samples

I  10000 test samples

I  Epoch 1/24

I  Epoch 1/24

I  Epoch 1/24
...
2019-02-09 17:48:19.763 VET Epoch 24/24
2019-02-09 17:48:19.763 VET Epoch 24/24
2019-02-09 17:48:19.763 VET Epoch 24/24

2019-02-09 17:49:27.756 VET Using TensorFlow backend.
2019-02-09 17:49:27.756 VET Test loss: 0.01727276442445009
2019-02-09 17:49:27.756 VET Test accuracy: 0.994

2019-02-09 17:49:27.756 VET Using TensorFlow backend.
2019-02-09 17:49:27.756 VET Test loss: 0.01727276442445009
2019-02-09 17:49:27.756 VET Test accuracy: 0.994

2019-02-09 17:49:27.756 VET Using TensorFlow backend.
2019-02-09 17:49:27.756 VET Test loss: 0.01727276442445009
2019-02-09 17:49:27.756 VET Test accuracy: 0.994
```
I don't understand if the model is running 24 epochs on each worker and each ends up with "3 models" or running 24 epochs between the 3 and combining the results.

Could someone please explain how horovod actually works and if there's something i could do to make it coordinate better and take advantage of resources??

How can I extract a single model at the end?

thanks in advance

**Comment by tgaddair** [comment]
Hey @josesuero, you're running 24 epochs with 1 model that's replicated across all 3 workers.  

In the `keras_mnist_advanced.py` example, the training data is split into `K` minibatches that are _randomly sampled_ from during training.  In every epoch, each of the `N` workers will sample `K / N` of the minibatches (so the more workers you have, the fewer minibatches per worker).

Each minibatch is processed _synchronously_ across all workers.  After the forward pass, the gradients are computed for the backward pass, then those gradients are _aggregated_ across all workers, and then applied to update the model weights.

What this means in effect is that:

1. Even though each of the workers operated on a different minibatch, the model is updated with the same gradients in each backward pass, therefore the models on all workers are exactly the same.
2. Each worker had a local minibatch size of M, but the effective minibatch size was actually M*N (the number of workers).

So what Horovod is doing is spreading the training data out across the workers, so less data needs to be processed per worker.  This is what results in linearly faster training as you add more workers. 

To extract the model at the end, you can just save any of the model replicas to disk (usually we use rank 0, but it's arbitrary).

**Comment by josesuero** [comment]
Thanks for your reply. I think I finally understand how this works. 

If I understood correctly when I put my own model all I have to do is make sure all workers have access to the same dataset and horvod will handle the splitting between them. 

Could you please confirm if I have to include anything else to the mpirun command or the YAML to make it "better" meaning more efficient for my when I do my own model

**Comment by abditag2** [comment]
@josesuero You can also take a look at the Horovod paper: https://arxiv.org/abs/1802.05799

**Comment by zoooo0820** [comment]
Hi, I'm confused about how to feed data when training on multi-nodes. For example, if I want to train a model on two servers (each has 4 GPUs, total 8GPUs),  as far as I know, I have to put the training scripts and datas on each server.  So here is the question, should I put data_all on both servers, or should I put data_half_1 on server1 and data_half_2 on server2?

**Comment by tgaddair** [comment]
@zoooo0820, all the data should be available to all the servers (it doesn't have to be physically there, it could be read from HDFS or any network mounted filesystem).  The reason being that after each epoch, you want to shuffle the data such that different workers will have different minibatches between epochs.

One way to handle reading and partitioning data across machines is to use a system like [Petastorm](https://github.com/uber/petastorm), but that's not the only way.  You can also simply put copies of the data on each machine, then shuffle the indices, and take a random batch on each worker at each step.

**Comment by zoooo0820** [comment]
@tgaddair  Thanks for your reply. That's quite clear.

**Comment by shunshunyin** [comment]
> @zoooo0820, all the data should be available to all the servers (it doesn't have to be physically there, it could be read from HDFS or any network mounted filesystem). The reason being that after each epoch, you want to shuffle the data such that different workers will have different minibatches between epochs.
> 
> One way to handle reading and partitioning data across machines is to use a system like [Petastorm](https://github.com/uber/petastorm), but that's not the only way. You can also simply put copies of the data on each machine, then shuffle the indices, and take a random batch on each worker at each step.

@tgaddair If we do not split and shuffle the data, do different workers process the same data?

**Comment by tgaddair** [comment]
@zoooo0820, yes, you need to handle data splitting somehow.  

The simplest way is to do what we do in the [tensorflow_mnist.py](https://github.com/horovod/horovod/blob/master/examples/tensorflow_mnist.py) example.  

Each of the N workers has the same data, but only performs 20000 / N steps with a random batch at each step. That way, we scale the amount of work with the number of workers, and by ensuring that each worker gets a different random batch at each step, also prevent them from processing the same data.

**Comment by jusonn** [comment]
@tgaddair How do I ensure that each worker gets a different random batch??
